src/main/java/com/faforever/client/map/generator/MapGeneratorService.java

```diff
@@ -0,0 +1,147 @@
+      }
+    }
+
+    seedGenerator = new Random();
+
+    customMapsDirectory = this.preferencesService.getPreferences().getForgedAlliance().getCustomMapsDirectory();
+    deleteGeneratedMaps();
+  }
+
+
+  private void deleteGeneratedMaps() {
+    logger.info("Deleting leftover generated maps...");
+
+    if (customMapsDirectory != null && customMapsDirectory.toFile().exists()) {
+      Arrays.stream(customMapsDirectory.toFile().listFiles())
+          .filter(File::isDirectory)
+          .filter(f -> GENERATED_MAP_PATTERN.matcher(f.getName()).matches())
+          .map(File::toPath)
+          .forEach(f -> noCatch(() -> FileUtils.deleteRecursively(f)));
+    }
+  }
+
+
+  public CompletableFuture<String> generateMap() {
+    return generateMap(GENERATOR_DEFAULT_VERSION, seedGenerator.nextLong());
+  }
+
+
+  //TODO: PROTECT AGAINST mapName INJECTION!!!!!
+  public CompletableFuture<String> generateMap(String mapName) {
+    Matcher matcher = GENERATED_MAP_PATTERN.matcher(mapName);
+    matcher.find();
+    return generateMap(matcher.group(1), Long.parseLong(matcher.group(2)));
+  }
+
+
+  public CompletableFuture<String> generateMap(String version, long seed) {
+
+    String generatorExecutableFileName = String.format(GENERATOR_EXECUTABLE_FILENAME, version);
+    File generatorExecutableFile = generatorExecutablePath.resolve(generatorExecutableFileName).toFile();
+
+//    notificationService.addNotification(new TransientNotification(i18n.get("game.mapGeneration.notification.title"), i18n.get("game.mapGeneration.notification.message", version, seed)));
+
+    CompletableFuture<Void> downloadGeneratorFuture;
+    if (!generatorExecutableFile.exists()) {
+      if (!VERSION_PATTERN.matcher(version).matches()) {
+        logger.error("Unsupported generator version: {}", version);
+        return CompletableFuture.supplyAsync(() -> {
+          throw new RuntimeException("Unsupported generator version: " + version);
+        });
+      }
+
+      logger.info("Downloading MapGenerator version: {}", version);
+      DownloadMapGeneratorTask downloadMapGeneratorTask = applicationContext.getBean(DownloadMapGeneratorTask.class);
+      downloadMapGeneratorTask.setVersion(version);
+      downloadGeneratorFuture = taskService.submitTask(downloadMapGeneratorTask).getFuture();
+    } else {
+      logger.info("Found MapGenerator version: {}", version);
+      downloadGeneratorFuture = CompletableFuture.completedFuture(null);
+    }
+
+    String mapFilename = String.format(MapGeneratorService.getGENERATED_MAP_NAME(), version, seed);
+
+    GenerateMapTask generateMapTask = applicationContext.getBean(GenerateMapTask.class);
+    generateMapTask.setVersion(version);
+    generateMapTask.setSeed(seed);
+    generateMapTask.setGeneratorExecutableFile(generatorExecutableFile);
+    generateMapTask.setMapFilename(mapFilename);
+
+    return downloadGeneratorFuture.thenApplyAsync((aVoid) -> {
```

Comment:
You mean .combineAsync() here, whatever you mean there are convience Methods for it

Reply:
I'd keep this for now as I want to return the map name (in theory I could return the map name in both of the other futures, but that is harder to understand for now)